rfcs/RFC-Uploads.md

```diff
@@ -0,0 +1,401 @@
+# User Uploads
+
+## For the cellxgene Data Portal
+
+**Authors:** [Trent Smith](mailto:trent.smith@chanzuckerberg.com)
+
+**Approvers:** [Arathi Mani](mailto:arathi.mani@chanzuckerberg.com), [Timmy Huang](mailto:thuang@chanzuckerberg.com),
+[Marcus Kinsella](mailto:mkinsella@chanzuckerberg.com), [Eduardo Lopez](mailto:elopez@chanzuckerberg.com), [Brian Raymore](mailto:braymor@chanzuckerberg.com)
+[Sara Gerber](mailto:sara.gerber@chanzuckerberg.com)
+
+## tl;dr
+
+A user can upload files of any size to Data Portal using either a shared link from their cloud provider or directly from their browser.
+
+## Glossary
+
+- CSP(Cloud Storage Provider) - Cloud storage services such as Dropbox or Google Drive.
+- Shared Link - A link from a CSP that gives the bearer of that link access to private data.
+- DP(Data Portal) - The cellxgene Data Portal.
+- Small File - a file < 5 GB in size.
+- Large File - a file > 5 GB in size.
+
+## Problem Statement | Background
+
+One of the primary goals of the DP is to allow users to upload their own files. These files consist of data sets as well
+as attestation files. These data sets can vary in size from a few MB to several GB. Users may want to upload from their local machine or their CSP. For users that want to upload from their local machine, they must be able to resume an
+upload in the event their connection is interrupted. For users that want to upload files from their CSP, they can provide a shared link from their CSP that is used to upload directly to S3.
+
+This specification does not cover the validation of the data or the process for a user to complete a submission.
+
+## Product Requirements
+
+1. A user can upload a file to DP using a shared link from their Dropbox.
+1. A user can upload a file to DP using a shared link from their Google Drive.
+1. A user can upload a file from their local machine using the DP browser.
+1. A user can resume an upload without completely restarting.
+1. The user must be authenticated before uploading.
+1. A user who has not yet agreed to the Portal policies must not be able to upload their file.
+
+### Nonfunctional requirements
+
+1. Cloud to Cloud uploads will retry if they fail. Retry at least 5 times with an exponential back off.
+1. A user can upload a single file up to 30GB in size.
```

Comment:
What is the purpose of the limit?     technical limitations,  avoiding abuse?   
If I look at the current datasets, one of the "original" datasets was ~17GB. 
It's conceivable that a legit dataset may exceed 30GB.

Reply:
This upper bound is a good size that we should be able to guarantee for our clients. There is no reason we can't extend it later. I wanted a solid number that we could test before declaring victory.

I thought the largest we are expecting was around 20GB in size. 30GB seems like a safe size to set as a goal that allows room to grow. 

FYI the S3 max is 5TB for a single file. I don't think we need to push to the limit for now.

Reply:
@mckinsel should validate this initial maximum.